tuned_lens/scripts/ingredients.py

```diff
@@ -0,0 +1,215 @@
+from functools import partial
+from hashlib import md5
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.distributed.fsdp import (
+    CPUOffload,
+    FullyShardedDataParallel,
+    MixedPrecision,
+)
+from typing import Optional, List
+from dataclasses import dataclass
+from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy
+from transformers import (
+    AutoConfig,
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    PreTrainedModel,
+    PreTrainedTokenizerBase,
+)
+from tuned_lens.data import (
+    chunk_and_tokenize,
+    compute_nats_to_bpb_ratio,
+)
+from tuned_lens.model_surgery import get_transformer_layers
+import pickle
+import shutil
+import torch as th
+import torch.distributed as dist
+
+from simple_parsing import field
+
+from tuned_lens.nn.lenses import TunedLens
+
+
+def local_rank() -> int:
+    """Get the local rank of the current process."""
+    return dist.get_rank() if dist.is_initialized() else 0
+
+
+def world_size() -> int:
+    """Get the world size from torch.distributed."""
+    return dist.get_world_size() if dist.is_initialized() else 1
+
+
+@dataclass
+class Data:
+    """Configuration for the dataset."""
+
+    name: List[str] = field(default_factory=lambda: ["the_pile", "all"], nargs="*")
+    """Name of dataset to use. Can either be a local .jsonl file or a name
+    suitable to be passed to the HuggingFace load_dataset function."""
+
+    split: str = "validation"
+    """Split of the dataset to use."""
+
+    text_column: str = "text"
+    """Column of the dataset containing text to run the model on."""
+
+    revision: Optional[str] = None
+
+    def load(self, tokenizer: PreTrainedTokenizerBase) -> tuple[Dataset, float]:
+        """Load the dataset, tokenize it and compute nats_to_bpb."""
+        print(f"Loading dataset '{' '.join(self.name)}'")
+
+        if len(self.name) == 1 and self.name[0].endswith(".jsonl"):
+            dataset = Dataset.from_json(self.name[0])
+            assert isinstance(dataset, Dataset)
+        else:
+            dataset = load_dataset(*self.name, split=self.split, revision=self.revision)
+            if not isinstance(dataset, (Dataset, DatasetDict)):
+                raise ValueError(
+                    "Only Dataset and DatasetDict instances are supported."
+                )
+
+        processed = chunk_and_tokenize(dataset, tokenizer, text_key=self.text_column)
+        nats_to_bpb = compute_nats_to_bpb_ratio(dataset, processed)
+
+        print(f"Using nats per token to bits per byte ratio: {nats_to_bpb}")
+
+        assert isinstance(processed, Dataset)
+        return processed, nats_to_bpb
+
+
+@dataclass
+class Model:
+    """Configuration for the model and tokenizer."""
+
+    name: str
+    """Name of model to use in the Huggingface Hub."""
+
+    no_cache: bool = field(action="store_true")
+    """Don't permanently cache the model on disk."""
+
+    random_model: bool = field(action="store_true")
```

Comment:
I think the `store_true` thing is not actually needed (like 80% confidence)

Reply:
No it is totally a thing https://github.com/lebrice/SimpleParsing/blob/1567bd19ef803c7703d76343584cc52feab0c00b/examples/custom_args/custom_args_example.py#L71

Reply:
Just to add a little from what I learned doing my refactor, I think you can technically get away with either specifying `store_true` or setting a default value, e.g. `random_model: bool = True`. I think the only difference is that specifying `store_true` means that the flag can't have a value specified for it in the command line, which setting a default value won't prevent. TLDR if you want to reproduce the original behavior of `add_argument` you want to use `action="store_true"`. Here's some test code I ran to show the behavior of the two options:

```
# with `action="store_true"`
> python tuned_lens/scripts/test.py --array
args: Options(array=True)

# with a default value of `False`
(env) ➜  tuned-lens git:(41-refactor-argument-parsers) ✗ python tuned_lens/scripts/test.py --array     
args: Options(array=False)

# trying to specify a value on the command line when using `store_true`:
python tuned_lens/scripts/test.py --array True
usage: test.py [-h] [--array]
test.py: error: unrecognized arguments: True

# when using `store_true` but not specifying the value in the command
> python tuned_lens/scripts/test.py
args: Options(array=False)
```